File: go-algorithm/leetcode/problem583.go
package leetcode

// 583. Delete Operation for Two Strings
// https://leetcode.com/problems/delete-operation-for-two-strings/
func minDistance(word1 string, word2 string) int {
	var min func(x, y int) int
	min = func(x, y int) int {
		if x < y {
			return x
		}
		return y
	}
	m, n := len(word1), len(word2)
	dp := make([][]int, m+1)
	for i := 0; i <= m; i++ {
		dp[i] = make([]int, n+1)
		dp[i][0] = i
	}
	for i := 0; i <= n; i++ {
		dp[0][i] = i
	}
	for i := 1; i <= m; i++ {
		for j := 1; j <= n; j++ {
			if word1[i-1] == word2[j-1] {
				dp[i][j] = dp[i-1][j-1]
			} else {
				dp[i][j] = min(min(dp[i-1][j]+1, dp[i][j-1]+1), dp[i-1][j-1]+2)
			}
		}
	}
	return dp[m][n]
}

File: java-algo/src/main/java/com/leetcode/Problem583.java
package com.leetcode;

/**
 * 583. Delete Operation for Two Strings
 * https://leetcode.com/problems/delete-operation-for-two-strings/
 */
public class Problem583 {
    public int minDistance(String word1, String word2) {
        int[][] memo = new int[word1.length() + 1][word2.length() + 1];
        return word1.length() + word2.length() - 2 * helper(word1, word2, word1.length(), word2.length(), memo);
    }

    private int helper(String a, String b, int m, int n, int[][] memo) {
        if (m == 0 || n == 0) {
            return 0;
        }
        if (memo[m][n] != 0) {
            return memo[m][n];
        }
        if (a.charAt(m - 1) == b.charAt(n - 1)) {
            memo[m][n] = helper(a, b, m - 1, n - 1, memo) + 1;
        } else {
            memo[m][n] = Math.max(helper(a, b, m, n - 1, memo), helper(a, b, m - 1, n, memo));
        }
        return memo[m][n];
    }
}
